src/Compilers/CSharp/Portable/Utilities/TypeUnionValueSet.cs

```diff
@@ -0,0 +1,510 @@
+    /// </summary>
+    internal sealed class TypeUnionValueSet : IValueSet
+    {
+        private readonly ConversionsBase _conversions;
+
+        /// <summary>
+        /// The set of types defining a union of types, instances of which could be in the value set.
+        /// </summary>
+        private readonly ImmutableArray<TypeSymbol> _typesInUnion;
+
+        /// <summary>
+        /// Root of a logical tree defining values contained in this value set.
+        /// 
+        /// If an instance of type cannot be an instance of any of the types in the union,
+        /// instances of that type are definitely not in the set. Otherwise, <see cref="_root"/>
+        /// defines if instances of a given type are in the set.
+        /// 
+        /// See <see cref="EvaluateNodeForInputValue"/> function.
+        /// If the tree evaluates to true for a given input type, instances of that type are definitely in the set.
+        /// If the tree evaluates to false for a given input type, instances of that type are definitely not in the set.
+        /// If the tree evaluates to 'null' (or an unknown result) for a given input type, instances of that type
+        /// might or might not be in the set, we cannot give a definitive answer.
+        /// 
+        /// </summary>
+        private readonly Node _root;
+
+        private bool? _lazyMightIncludeNonNull;
+        private bool? _lazyIncludesNull;
+
+        private TypeUnionValueSet(
+            ImmutableArray<TypeSymbol> typesInUnion,
+            Node root,
+            ConversionsBase conversions)
+        {
+            Debug.Assert(!typesInUnion.Any(t => t.IsNullableType()));
+            Debug.Assert(typesInUnion.Distinct().Length == typesInUnion.Length);
+            _typesInUnion = typesInUnion;
+            _root = root;
+            _conversions = conversions;
+        }
+
+        internal static TypeUnionValueSet AllValues(ImmutableArray<TypeSymbol> typesInUnion, ConversionsBase conversions)
+        {
+            return new TypeUnionValueSet(typesInUnion, IsTrueNode.Instance, conversions);
+        }
+
+        internal static TypeUnionValueSet FromTypeMatch(ImmutableArray<TypeSymbol> typesInUnion, TypeSymbol type, ConversionsBase conversions, ref CompoundUseSiteInfo<AssemblySymbol> useSiteInfo)
+        {
+            if (AnyTypeFromUnionMightMatch(typesInUnion, type, conversions, ref useSiteInfo))
+            {
+                return new TypeUnionValueSet(typesInUnion, new IsTypeNode(type), conversions);
+            }
+
+            // An empty set
+            return new TypeUnionValueSet(typesInUnion, IsFalseNode.Instance, conversions);
+        }
+
+        private static bool AnyTypeFromUnionMightMatch(ImmutableArray<TypeSymbol> typesInUnion, TypeSymbol type, ConversionsBase conversions, ref CompoundUseSiteInfo<AssemblySymbol> useSiteInfo)
+        {
+            Debug.Assert(!typesInUnion.IsEmpty);
```

Comment:
I didn't follow how it is that we are certain of this. The other two assertions I understand, but, I didn't follow why it is not possible for a pattern matching against an empty union to make it here.